### ISSUE TITLE
QueryMap not being encoded when value starts with '{'

### ISSUE THREAD

**Issue by seno5979** [OpenFeign/feign#694]
Hi,

When using QueryMap(encoded=true) on my feignClient, part of the query map value is not being encoded due to below logic (the value start with '{'), Is it a bug ? 
```java
public final class RequestTemplate implements Serializable {
...
...
 private static String encodeIfNotVariable(String in) {
    if (in == null || in.indexOf('{') == 0) {
      return in;
    }
    return urlEncode(in);
  }
```
https://github.com/OpenFeign/feign/blob/master/core/src/main/java/feign/RequestTemplate.java#L389

**Comment by kdavisk6** [comment]
Do you have an example interface demonstrating the issue?  At first glance, this is correct and you can define parameters in the interface definition like so:

```java
@RequestLine("GET /api/users/{id}?param={param}")
public Users getUsers(String id, @QueryMap Map<String, String> parameters);
```

So we don't want to encode variable names.

**Comment by kdavisk6** [comment]
Also, you have specified `QueryMap(encoded = true)`.  This indicated to `Feign` that the values are already encoded and will not perform any encoding.

**Comment by kdavisk6** [comment]
In looking into other issues, I have found out why you are experiencing it.  If a parameter contains a start or end brace `{` `}`, the value is considered a variable, which makes it subject to the same processing as other variables.  Feign will attempt to resolve the variable before sending the request.  

Currently, when a variable is defined but does not contain a value, the variable value is considered a `literal` value.  This is why the information is not being encoded.  In addition to my previous comment, the `encoded` value is to indicate to that the values are already encoded.  My suggestion is to remove the `encoded = true` from the `@QueryMap` annotation and then use two braces `{{ item }}`.  By wrapping the parameter in double braces, the value will be seen as a braced literal, resulting in `{ item }` being used.  Give that a try.

**Comment by kdavisk6** [comment]
See #778

**Comment by kdavisk6** [comment]
Closed via #778
